[processor/filter] Add telemetry for dropped metrics, logs, and spans

MacroPower opened the pull request:
**Description:**

Adds telemetry for metrics, logs, and spans that were intentionally dropped via a `filterprocessor`. Specifically, the following metrics are added:

`otelcol_processor_filter_datapoints_filtered`
`otelcol_processor_filter_logs_filtered`
`otelcol_processor_filter_spans_filtered`

Please let me know any feedback/thoughts on the naming or anything else!

**Link to tracking Issue:** #13169

**Testing:** I've used batchprocessor as an example for a couple of tests, Filter*ProcessorTelemetryWithOC. I kept the wrapping code so that OTEL versions can be easily added when that is ready in contrib. Overall the tests are not super comprehensive and I could improve them if needed, but as-is they were helpful for debugging.

<details>
  <summary><i>Additionally, here's some stuff you can use for manually testing.</i></summary>

There might be a better way to do this, but I just used hostmetrics, filelog, and [this article from honeycomb](https://www.honeycomb.io/blog/test-span-opentelemetry-collector) with otlp/http.

Note, this should be run from the root of the contrib repo.

Add/overwrite `local/config.yaml`, `local/span.json`, and run:

```bash
mkdir -p local

cat >local/config.yaml <<EOL
receivers:
  hostmetrics:
    collection_interval: 30s
    initial_delay: 1s
    scrapers:
      load:
  filelog:
    include:
      ## echo '{"timestamp":"2023-12-18 12:00:00","msg":"foo"}' >> /tmp/otel-test.log
      ## echo '{"timestamp":"2023-12-18 12:00:00","msg":"bar"}' >> /tmp/otel-test.log
      ## echo '{"timestamp":"2023-12-18 12:00:00","msg":"baz"}' >> /tmp/otel-test.log
      - /tmp/otel-test.log
    operators:
      - type: json_parser
        timestamp:
          parse_from: attributes.timestamp
          layout: "%Y-%m-%d %H:%M:%S"
  otlp:
    protocols:
      ## curl -i http://localhost:4318/v1/traces -X POST -H "Content-Type: application/json" -d @local/span.json
      http:

processors:
  filter/test:
    metrics:
      metric:
        # Should drop 2 of the 3 metrics, 5m average remains
        - 'name=="system.cpu.load_average.1m"'
        - 'name=="system.cpu.load_average.15m"'
    logs:
      log_record:
        # Should filter out "bar" and "baz"
        - 'IsMatch(body, ".*ba.*")'
    traces:
      span:
        # Should drop 1 of the 2 spans
        - 'name == "foobar"'

exporters:
  debug:
    verbosity: detailed
    sampling_initial: 5
    sampling_thereafter: 200

service:
  extensions: []
  pipelines:
    metrics:
      receivers: [hostmetrics]
      processors: [filter/test]
      exporters: [debug]
    logs:
      receivers: [filelog]
      processors: [filter/test]
      exporters: [debug]
    traces:
      receivers: [otlp]
      processors: [filter/test]
      exporters: [debug]

  telemetry:
    logs:
      level: debug
    metrics:
      level: detailed
      address: 0.0.0.0:8888
EOL

cat >local/span.json <<EOL
{
  "resourceSpans": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "test-with-curl"
            }
          }
        ]
      },
      "scopeSpans": [
        {
          "scope": {
            "name": "manual-test"
          },
          "spans": [
            {
              "traceId": "71699b6fe85982c7c8995ea3d9c95df2",
              "spanId": "3c191d03fa8be065",
              "name": "spanitron",
              "kind": 2,
              "droppedAttributesCount": 0,
              "events": [],
              "droppedEventsCount": 0,
              "status": {
                "code": 1
              }
            },
            {
              "traceId": "71699b6fe85982c7c8995ea3d9c95df2",
              "spanId": "2f357b34d32f77b4",
              "name": "foobar",
              "kind": 2,
              "droppedAttributesCount": 0,
              "events": [],
              "droppedEventsCount": 0,
              "status": {
                "code": 1
              }
            }
          ]
        }
      ]
    }
  ]
}
EOL

make run
```

Send some data to the receivers:

```bash
# Write some logs
echo '{"timestamp":"2023-12-18 12:00:00","msg":"foo"}' >> /tmp/otel-test.log
echo '{"timestamp":"2023-12-18 12:00:00","msg":"bar"}' >> /tmp/otel-test.log
echo '{"timestamp":"2023-12-18 12:00:00","msg":"baz"}' >> /tmp/otel-test.log

# Write some spans
curl -i http://localhost:4318/v1/traces -X POST -H "Content-Type: application/json" -d @local/span.json
```

Check the results:

```console
$ curl http://localhost:8888/metrics | grep filtered
# HELP otelcol_processor_filter_datapoints_filtered Number of metric data points dropped by the filter processor
# TYPE otelcol_processor_filter_datapoints_filtered counter
otelcol_processor_filter_datapoints_filtered{filter="filter/test",service_instance_id="a99d9078-548b-425f-8466-3e9e2e9bf3b1",service_name="otelcontribcol",service_version="0.91.0-dev"} 2
# HELP otelcol_processor_filter_logs_filtered Number of logs dropped by the filter processor
# TYPE otelcol_processor_filter_logs_filtered counter
otelcol_processor_filter_logs_filtered{filter="filter/test",service_instance_id="a99d9078-548b-425f-8466-3e9e2e9bf3b1",service_name="otelcontribcol",service_version="0.91.0-dev"} 2
# HELP otelcol_processor_filter_spans_filtered Number of spans dropped by the filter processor
# TYPE otelcol_processor_filter_spans_filtered counter
otelcol_processor_filter_spans_filtered{filter="filter/test",service_instance_id="a99d9078-548b-425f-8466-3e9e2e9bf3b1",service_name="otelcontribcol",service_version="0.91.0-dev"} 1
```

</details>

**Documentation:** I do not believe we document telemetry exposed by components, but I could add this if needed.

MacroPower committed:
Add initial telemetry for filterprocessor metrics

atoulme commented on processor/filterprocessor/factory.go:
```diff
@@ -15,6 +15,11 @@ import (
 
+const (
+	// The value of "type" key in configuration.
+	typeStr = "filter"
```
You can use metadata.Type instead, it is generated from metadata.yaml.

TylerHelmuth commented on processor/filterprocessor/metrics.go:
```diff
@@ -36,6 +38,13 @@ func newFilterMetricProcessor(set component.TelemetrySettings, cfg *Config) (*fi
+
+	fpt, err := newfilterProcessorTelemetry(set)
+	if err != nil {
+		return nil, fmt.Errorf("error creating batch processor telemetry: %w", err)
```
```suggestion
		return nil, fmt.Errorf("error creating filter processor telemetry: %w", err)
```

TylerHelmuth commented on processor/filterprocessor/telemetry.go:
```diff
@@ -0,0 +1,127 @@
+
+var (
+	processorTagKey      = tag.MustNewKey(processorKey)
+	statMetricsFiltered  = stats.Int64("metrics_filtered", "Number of metrics dropped by the filter processor", stats.UnitDimensionless)
+	statMetricsProcessed = stats.Int64("metrics_processed", "Number of metrics processed by the filter processor", stats.UnitDimensionless)
```
I think we should use `.` instead of `_` to namespace the metric names (some other components don't do this yet bc they were created before that naming was determined but I believe we're in the middle of updating them).

TylerHelmuth commented on processor/filterprocessor/metrics.go:
```diff
@@ -168,6 +179,7 @@ func (fmp *filterMetricProcessor) processMetrics(ctx context.Context, md pmetric
 	if md.ResourceMetrics().Len() == 0 {
 		return md, processorhelper.ErrSkipProcessingData
 	}
+	fmp.telemetry.record(int64(totalMetrics-md.ResourceMetrics().Len()), int64(totalMetrics))
```
This is recording the number of ResourcesMetrics dropped, but not the number of metrics themselves.  We'll need to move this recording somewhere else or update `totalMetrics` to count the number or metrics inside the `ResourceMetrics` slice.

MacroPower replied:
I believe I've fixed this, initially I did not understand the distinction between `ResourceMetrics`, `Metrics`, and `DataPoints`.

MacroPower committed:
Fix filterprocessor telemetry naming

MacroPower committed:
Fix filterprocessor dropped telemetry, remove total metrics

MacroPower committed:
Create FilterMetricProcessor with processor.CreateSettings

This is required so that we can access metadata about the processor for telemetry tags/labels.

MacroPower committed:
Add support for filterprocessor telemetry from logs, traces

MacroPower committed:
Add filterprocessor log, trace telemetry

MacroPower committed:
Set filterprocessor typeStr from metadata.Type

MacroPower committed:
Remove optional OTEL metrics from filterprocessor telemetry

This feature flag is not available for contrib packages.

MacroPower committed:
Merge branch 'main' into main

MacroPower committed:
Fix filterprocessor OC telemetry name, aggregation

MacroPower committed:
Add basic filterprocessor telemetry tests

MacroPower committed:
Add changelog entry for new filterprocessor telemetry

MacroPower commented:
It took a while but I think this is ready for review for real. Thanks for the feedback on my draft. Also wanted to say that Contribfest was really awesome! I had a ton of run working with everyone, and I'm also glad to finally work on the collector a bit considering how frequently I use it day-to-day. 🙂

MacroPower committed:
Fix filterprocessor formatting/naming per linter

TylerHelmuth commented on processor/filterprocessor/telemetry.go:
```diff
@@ -0,0 +1,101 @@
+var (
+	typeStr             = metadata.Type
+	processorTagKey     = tag.MustNewKey(typeStr)
+	statMetricsFiltered = stats.Int64("metrics.filtered", "Number of metrics dropped by the filter processor", stats.UnitDimensionless)
```
Lets count datapoints instead of metrics using `md.DataPointCount`

TylerHelmuth commented:
@MacroPower can you share some details on the new metrics in action from your manual tests?

MacroPower committed:
Merge remote-tracking branch 'upstream/main'

# Conflicts:
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum

MacroPower committed:
Count filterprocessor metric data points dropped

TylerHelmuth commented on processor/filterprocessor/telemetry.go:
```diff
@@ -0,0 +1,101 @@
+var (
+	typeStr                      = metadata.Type
+	processorTagKey              = tag.MustNewKey(typeStr)
+	statMetricDataPointsFiltered = stats.Int64("metrics.filtered", "Number of metric data points dropped by the filter processor", stats.UnitDimensionless)
```
```suggestion
	statMetricDataPointsFiltered = stats.Int64("datapoints.filtered", "Number of metric data points dropped by the filter processor", stats.UnitDimensionless)
```

MacroPower replied:
Sure, and apologies for the wait, EOY crunch has been brutal.

I considered changing this name but was a bit worried that `datapoints` would be unclear, like normally you'd have logs, metrics, and traces, *OR* the actual OTTL contexts, but not both.

Do you think it'd be a better idea to add individual metrics for all OTTL contexts? i.e., span, spanevent, metric, datapoint, and log_record?

TylerHelmuth replied:
Since we're counting the actual data dropped (whether by OTTL or by the legacy method) I think we should omit the OTTL context for now. The reason I vote for the metric to be namespaced with `datapoints` is because we are using the `DataPointCount` function to count datapoints, not metrics.

I think in the future, in a separate PR, we could add metrics for `metrics.filtered` and `span_events.filtered`.

MacroPower replied:
That's fair, done!

TylerHelmuth reviewed (commented):
Thanks for sticking with this!

MacroPower committed:
Rename metrics.filtered to datapoints.filtered

MacroPower committed:
Merge branch 'main' into main

MacroPower commented:
> @MacroPower can you share some details on the new metrics in action from your manual tests?

@TylerHelmuth I edited my PR description to include some stuff that can be used for manual testing. It's not the best and there's probably a better way I imagine, but hopefully it's useful regardless. 🙂

TylerHelmuth reviewed (approved):
@MacroPower thanks for the testing details, I was able to confirm locally as well.  Take care of the conflicts and we'll be good to merge.

MacroPower committed:
Merge remote-tracking branch 'upstream/main'

# Conflicts:
#	processor/filterprocessor/go.mod
#	processor/filterprocessor/go.sum

MacroPower committed:
Merge branch 'main' into main

TylerHelmuth committed:
Merge branch 'main' into main

TylerHelmuth committed:
Merge branch 'main' into main

TylerHelmuth committed:
Merge branch 'main' into main

vainiusd commented:
Hi, Noticed this only on the v0.92.0 release notes so sorry for the late feedback.
And also thanks for adding this!

The new metric names stand out a bit compared to others:
```
otelcol_receiver_refused_signal
otelcol_processor_refused_signal
otelcol_processor_filter_signal_filtered
otelcol_processor_dropped_signal
otelcol_exporter_enqueue_failed_signal
otelcol_exporter_send_failed_signal
```

Maybe it would be worth renaming to `otelcol_processor_filter_filtered_signal`?
